### ISSUE TITLE
3.11 regression: traceback.format_list raises UnicodeDecodeError in certain scenarios

### ISSUE THREAD

**Issue by sebastinas** [python/cpython#98744]
# Bug report

Take the following piece of code:
```py3
import sys, traceback
try:
    ｗｉｄｔｈ
except:
    _, _, tb = sys.exc_info()

tblist = traceback.extract_tb(tb)
print(traceback.format_list(tblist))
```

With Python 3.10 and earlier versions, executing a file with this code produces:
```
['  File "/tmp/test.py", line 3, in <module>\n    ｗｉｄｔｈ\n']
```

With Python 3.11, a `UnicodeDecodeError` is raised instead:
```
Traceback (most recent call last):
  File "/tmp/test.py", line 9, in <module>
    print(traceback.format_list(tblist))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/traceback.py", line 41, in format_list
    return StackSummary.from_list(extracted_list).format()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/traceback.py", line 531, in format
    formatted_frame = self.format_frame_summary(frame_summary)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/traceback.py", line 478, in format_frame_summary
    colno = _byte_offset_to_character_offset(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/traceback.py", line 566, in _byte_offset_to_character_offset
    return len(as_utf8[:offset + 1].decode("utf-8"))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xef in position 4: unexpected end of data
```

# Your environment

- CPython versions tested on: 3.10.8 and 3.11.0
- Operating system and architecture: Debian unstable, amd64

**Comment by sebastinas** [comment]
Current main (d578aaea6257458c199328100cbb5af64c6a043e) also has the same issue.

**Comment by serhiy-storchaka** [comment]
cc @ammaraskar, @pablogsal, @isidentical

**Comment by mdboom** [comment]
Also cc @iritkatriel (related to source positions -- in this case a source position in the middle of a UTF-8 codepoint).

**Comment by isidentical** [comment]
It seems like we are not mirroring the behaviour from the C implementation, which uses `errors="replace"` instead of the default `strict`. I'll try to send a PR soon. Thanks for the report @sebastinas!

**Comment by mdboom** [comment]
@isidentical: I was just looking at this as well, and there is also an off-by-one difference between this and the C implementation.

**Comment by isidentical** [comment]
@mdboom you are right! That is also an issue (in the C Implementation, we do the length increment during bounds check time but apparently in the Python implementation it is done afterwards). Both should now be resolved with #98824. Thanks for catching it.

**Comment by isidentical** [comment]
Should be fixed now, thanks everyone!
